// @shopify/cli-kit@3.94.3/dist/public/node/themes/theme-manager.js
import { fetchTheme, findDevelopmentThemeByName, themeCreate } from './api.js';
import { DEVELOPMENT_THEME_ROLE } from './utils.js';
import { generateThemeName } from '../../../private/node/themes/generate-theme-name.js';
import { BugError } from '../error.js';
export class ThemeManager {
    constructor(adminSession) {
        this.adminSession = adminSession;
    }
    async findOrCreate(name, role) {
        let theme = await this.fetch(name, role);
        if (theme) {
            this.setTheme(theme.id.toString());
        }
        else {
            theme = await this.create(role, name);
        }
        return theme;
    }
    async fetch(name, role) {
        if (!this.themeId && !name) {
            return;
        }
        const theme = name && role === DEVELOPMENT_THEME_ROLE
            ? await findDevelopmentThemeByName(name, this.adminSession)
            : await fetchTheme(parseInt(this.themeId, 10), this.adminSession);
        if (!theme) {
            this.removeTheme();
        }
        return theme;
    }
    generateThemeName(context) {
        return generateThemeName(context);
    }
    async create(themeRole, themeName) {
        const name = themeName ?? generateThemeName(this.context);
        const role = themeRole ?? DEVELOPMENT_THEME_ROLE;
        const theme = await themeCreate({
            name,
            role,
        }, this.adminSession);
        if (!theme) {
            throw new BugError(`Could not create theme with name "${name}" and role "${role}"`);
        }
        this.setTheme(theme.id.toString());
        return theme;
    }
}
