// @shopify/cli-kit@3.94.3/dist/private/node/api/graphql.js
import { GraphQLClientError, sanitizedHeadersOutput } from './headers.js';
import { sanitizeURL } from './urls.js';
import { stringifyMessage, outputContent, outputToken, outputDebug } from '../../../public/node/output.js';
import { AbortError } from '../../../public/node/error.js';
import { ClientError } from 'graphql-request';
export function debugLogRequestInfo(api, query, url, variables, headers = {}) {
    outputDebug(outputContent `Sending ${outputToken.json(api)} GraphQL request:
  ${outputToken.raw(query.toString().trim())}
${variables ? `\nWith variables:\n${sanitizeVariables(variables)}\n` : ''}
With request headers:
${sanitizedHeadersOutput(headers)}\n
to ${sanitizeURL(url)}`);
}
export function sanitizeVariables(variables) {
    const result = { ...variables };
    const sensitiveKeys = ['apiKey', 'serialized_script'];
    const sanitizedResult = sanitizeDeepVariables(result, sensitiveKeys);
    return JSON.stringify(sanitizedResult, null, 2);
}
function sanitizeDeepVariables(value, sensitiveKeys) {
    // Checking for JSON
    if (typeof value === 'string') {
        try {
            const parsed = JSON.parse(value);
            if (typeof parsed === 'object' && parsed !== null) {
                const sanitized = sanitizeDeepVariables(parsed, sensitiveKeys);
                return JSON.stringify(sanitized, null);
            }
            // eslint-disable-next-line no-catch-all/no-catch-all
        }
        catch (error) {
            return value;
        }
    }
    if (typeof value !== 'object' || value === null) {
        return value;
    }
    if (Array.isArray(value)) {
        return value.map((item) => sanitizeDeepVariables(item, sensitiveKeys));
    }
    const result = {};
    for (const [key, val] of Object.entries(value)) {
        if (sensitiveKeys.includes(key) && typeof val === 'string') {
            result[key] = '*****';
            continue;
        }
        result[key] = sanitizeDeepVariables(val, sensitiveKeys);
    }
    return result;
}
/**
 * Extracts human-readable error messages from a GraphQL errors array.
 *
 * Some APIs (e.g. App Management) return structured errors nested inside
 * `extensions.app_errors.errors[].message`. When those are present, we extract
 * them so the CLI displays a clean message instead of a raw JSON dump.
 * Falls back to each error's top-level `message` field, and ultimately to
 * the full JSON representation if no messages can be extracted.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function extractGraphQLErrorMessages(errors) {
    if (!errors || errors.length === 0)
        return undefined;
    const messages = [];
    for (const error of errors) {
        // Check for nested app_errors (App Management API pattern)
        const appErrors = error?.extensions?.app_errors?.errors;
        if (Array.isArray(appErrors) && appErrors.length > 0) {
            const appMessages = [];
            for (const appError of appErrors) {
                const friendlyMessage = friendlyAppErrorMessage(appError);
                if (friendlyMessage) {
                    appMessages.push(friendlyMessage);
                }
                else if (typeof appError?.message === 'string') {
                    appMessages.push(appError.message);
                }
            }
            // Fall back to top-level error message if no app_error messages were extracted
            if (appMessages.length > 0) {
                messages.push(...appMessages);
            }
            else if (typeof error?.message === 'string') {
                messages.push(error.message);
            }
        }
        // Fall back to top-level error message
        else if (typeof error?.message === 'string') {
            messages.push(error.message);
        }
    }
    return messages.length > 0 ? messages.join('\n') : undefined;
}
/**
 * Maps known app_errors categories to user-friendly messages.
 * Returns undefined if the error doesn't match a known pattern.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function friendlyAppErrorMessage(appError) {
    if (appError?.category === 'access_denied') {
        return "You don't have the necessary permissions to perform this action. Check that you're using the correct account or token.";
    }
    return undefined;
}
export function errorHandler(api) {
    return (error, requestId) => {
        if (error instanceof ClientError) {
            const { status } = error.response;
            const extractedMessages = extractGraphQLErrorMessages(error.response.errors);
            let errorMessage;
            if (extractedMessages && status < 500) {
                errorMessage = extractedMessages;
            }
            else if (extractedMessages && status >= 500) {
                errorMessage = `The ${api} GraphQL API responded with HTTP status ${status}: ${extractedMessages}`;
            }
            else {
                errorMessage = stringifyMessage(outputContent `
The ${outputToken.raw(api)} GraphQL API responded unsuccessfully with${status === 200 ? '' : ` the HTTP status ${status} and`} errors:

${outputToken.json(error.response.errors)}
        `);
            }
            if (requestId) {
                errorMessage += `\n\nRequest ID: ${requestId}`;
            }
            let mappedError;
            if (status < 500) {
                mappedError = new GraphQLClientError(errorMessage, status, error.response.errors);
            }
            else {
                mappedError = new AbortError(errorMessage);
            }
            mappedError.stack = error.stack;
            return mappedError;
        }
        else {
            return error;
        }
    };
}
