// @shopify/cli-kit@3.94.3/dist/private/node/analytics.js
import { getLastSeenAuthMethod } from './session.js';
import { getAutoUpgradeEnabled } from './conf-store.js';
import { hashString } from '../../public/node/crypto.js';
import { getPackageManager, packageManagerFromUserAgent } from '../../public/node/node-package-manager.js';
import * as metadata from '../../public/node/metadata.js';
import { platformAndArch } from '../../public/node/os.js';
import { ciPlatform, cloudEnvironment, macAddress } from '../../public/node/context/local.js';
import { cwd } from '../../public/node/path.js';
import { currentProcessIsGlobal, inferPackageManagerForGlobalCLI } from '../../public/node/is-global.js';
import { isWsl } from '../../public/node/system.js';
export async function startAnalytics({ commandContent, args, currentTime = new Date().getTime(), commandClass, }) {
    let startCommand = commandContent.command;
    if (commandClass && Object.prototype.hasOwnProperty.call(commandClass, 'analyticsNameOverride')) {
        startCommand = commandClass.analyticsNameOverride() ?? commandContent.command;
    }
    let pluginName = commandClass?.plugin?.name;
    if (commandClass && 'customPluginName' in commandClass) {
        pluginName = commandClass.customPluginName;
    }
    await metadata.addSensitiveMetadata(() => ({
        commandStartOptions: {
            startTime: currentTime,
            startCommand,
            startArgs: args,
        },
    }));
    await metadata.addPublicMetadata(() => ({
        cmd_all_launcher: packageManagerFromUserAgent(),
        cmd_all_alias_used: commandContent.alias,
        cmd_all_topic: commandContent.topic,
        cmd_all_plugin: pluginName,
        cmd_all_force: flagIncluded('force', commandClass) ? args.includes('--force') : undefined,
    }));
}
export async function getEnvironmentData(config) {
    const ciplatform = ciPlatform();
    const pluginNames = getPluginNames(config);
    const shopifyPlugins = pluginNames.filter((plugin) => plugin.startsWith('@shopify/'));
    const { platform, arch } = platformAndArch();
    return {
        uname: `${platform} ${arch}`,
        env_ci: ciplatform.isCI,
        env_ci_platform: ciplatform.name,
        env_plugin_installed_any_custom: pluginNames.length !== shopifyPlugins.length,
        env_plugin_installed_shopify: JSON.stringify(shopifyPlugins),
        env_shell: config.shell,
        env_web_ide: cloudEnvironment().editor ? cloudEnvironment().platform : undefined,
        env_device_id: hashString(await macAddress()),
        env_cloud: cloudEnvironment().platform,
        env_package_manager: await getPackageManager(cwd()),
        env_install_package_manager: inferPackageManagerForGlobalCLI(),
        env_is_global: currentProcessIsGlobal(),
        env_auth_method: await getLastSeenAuthMethod(),
        env_is_wsl: await isWsl(),
        env_build_repository: process.env.SHOPIFY_CLI_BUILD_REPO ?? 'unknown',
        env_auto_upgrade_enabled: getAutoUpgradeEnabled() ?? null,
    };
}
export async function getSensitiveEnvironmentData(config) {
    return {
        env_plugin_installed_all: JSON.stringify(getPluginNames(config)),
        env_shopify_variables: JSON.stringify(getShopifyEnvironmentVariables()),
    };
}
function getShopifyEnvironmentVariables() {
    // Agent callers can identify themselves today via SHOPIFY_* environment
    // variables. The current contract is intentionally lightweight and is kept in
    // the sensitive payload until we prove which dimensions deserve first-class
    // Monorail fields, e.g. SHOPIFY_CLI_AGENT, SHOPIFY_CLI_AGENT_VERSION,
    // SHOPIFY_CLI_AGENT_RUN_ID, SHOPIFY_CLI_AGENT_SESSION_ID, and
    // SHOPIFY_CLI_AGENT_PROVIDER.
    return Object.fromEntries(Object.entries(process.env).filter(([key]) => key.startsWith('SHOPIFY_')));
}
function getPluginNames(config) {
    const pluginNames = [...config.plugins.keys()];
    return pluginNames.sort().filter((plugin) => !plugin.startsWith('@oclif/'));
}
function flagIncluded(flag, commandClass) {
    if (!commandClass)
        return false;
    const commandFlags = commandClass.flags ?? {};
    return Object.keys(commandFlags).includes(flag);
}
