// @shopify/cli-kit@3.94.3/dist/private/node/ui.js
import { output } from './output.js';
import { isUnitTest } from '../../public/node/context/local.js';
import { treeKill } from '../../public/node/tree-kill.js';
import React, { createContext, useCallback, useContext, useEffect, useState } from 'react';
import { render as inkRender, useApp } from 'ink';
import { EventEmitter } from 'events';
const CompletionContext = createContext(null);
/**
 * Signal that the current Ink tree is done. Must be called within an
 * InkLifecycleRoot — throws if the provider is missing so lifecycle
 * bugs surface immediately instead of silently hanging.
 */
export function useComplete() {
    const complete = useContext(CompletionContext);
    if (!complete) {
        throw new Error('useComplete() called outside InkLifecycleRoot');
    }
    return complete;
}
/**
 * Root wrapper for Ink trees. Owns the single `exit()` call site — children
 * signal completion via `useComplete()`, which sets state here. The `useEffect`
 * fires post-render, guaranteeing all batched state updates have been flushed
 * before the tree is torn down.
 */
export function InkLifecycleRoot({ children }) {
    const { exit } = useApp();
    const [exitResult, setExitResult] = useState(null);
    const complete = useCallback((error) => {
        setExitResult({ error });
    }, []);
    useEffect(() => {
        if (exitResult !== null) {
            exit(exitResult.error);
        }
    }, [exitResult, exit]);
    return React.createElement(CompletionContext.Provider, { value: complete }, children);
}
export function renderOnce(element, { logLevel = 'info', renderOptions }) {
    const { output: renderedString, unmount } = renderString(element, renderOptions);
    if (renderedString) {
        output(renderedString, logLevel);
    }
    unmount();
    return renderedString;
}
export async function render(element, options) {
    const { waitUntilExit } = inkRender(React.createElement(InkLifecycleRoot, null, element), options);
    await waitUntilExit();
}
export class Stdout extends EventEmitter {
    constructor(options) {
        super();
        this.frames = [];
        this.write = (frame) => {
            this.frames.push(frame);
            // Ink writes `this.lastOutput + '\n'` to stdout during unmount when
            // running in a CI environment (detected via `is-in-ci`).  In debug
            // mode (which tests use), `lastOutput` is never updated, so the write
            // is just '\n', clobbering the last real rendered frame.  Skip it so
            // that `lastFrame()` keeps returning the final rendered content.
            if (frame !== '\n') {
                this._lastFrame = frame;
            }
        };
        this.lastFrame = () => {
            return this._lastFrame;
        };
        this.columns = options.columns ?? 80;
        this.rows = options.rows ?? 80;
    }
}
const renderString = (element, renderOptions) => {
    const columns = isUnitTest() ? 80 : process.stdout.columns;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const stdout = renderOptions?.stdout ?? new Stdout({ columns });
    const instance = inkRender(element, {
        stdout,
        debug: true,
        exitOnCtrlC: false,
        patchConsole: false,
    });
    return {
        output: stdout.lastFrame(),
        unmount: instance.unmount,
    };
};
export function handleCtrlC(input, key, exit = () => {
    treeKill(process.pid, 'SIGINT');
}) {
    if (input === 'c' && key.ctrl) {
        // Exceptions thrown in hooks aren't caught by our errorHandler.
        exit();
    }
}
